src/box/lua/net_box.lua

```diff
@@ -942,6 +962,99 @@ local console_mt = {
     __metatable = false
 }
 
+local weak_value_mt = {__mode = 'v'}
+
+local stream_index_mt = {
+    __index = function(self, key)
+        return self._src[key]
+    end
+}
+
+-- Create stream space index, which is same as connection space
+-- index, but have non zero stream ID.
+local function stream_wrap_index(stream_id, src)
+    return setmetatable({
+        _stream_id = stream_id,
+        _src = src,
+    }, stream_index_mt)
+end
+
+-- Metatable for stream space indexes. When stream space being
+-- created there are no indexes in it. When accessing the space
+-- index, we look for corresponding space index in corresponding
+-- connection space. If it is found we create same index for the
+-- stream space but with corresponding stream ID.
+local stream_indexes_mt = {
+    __index = function(self, key)
+        local _space = self._space
+        local src = _space._src.index[key]
+        if not src then
+            return nil
+        end
+        local res = stream_wrap_index(_space._stream_id, src)
+        self[key] = res
+        return res
+    end
+}
+
+local stream_space_mt = {
+    __index = function(self, key)
+        return self._src[key]
+    end
+}
+
+-- Create stream space, which is same as connection space,
+-- but have non zero stream ID.
+local function stream_wrap_space(stream, src)
+    local res = setmetatable({
+        _stream_id = stream._stream_id,
+        _src = src,
+        index = setmetatable({
+            _space = nil,
+        }, stream_indexes_mt)
+    }, stream_space_mt)
+    res.index._space = res
+    return res
+end
+
+-- Metatable for stream spaces. When stream being created
+-- there are no spaces in it. When user try to access some
+-- space in stream, we look for corresponding space in connection.
+-- If it is found we create same space for the stream but with
+-- corresponding stream ID. Also we set `_space_was_accessed` to
+-- flag, to mark that fact that there was access for some space to
+-- this stream. Later when `schema_version` reloaded we reload all
+-- spaces for all streams for which there was access to any spaces.
+local stream_spaces_mt
+stream_spaces_mt = {
+    __index = function(self, key)
+        local stream = self._stream
+        local src = stream._conn.space[key]
+        if not src then
+            return nil
+        end
+        if not stream._space_was_touched then
+            -- Subscribe on reload on demand.
+            local weak = stream._weak
+            local conn = stream._conn
+            local function on_schema_reload()
+                local stream = weak.stream
+                stream.space = setmetatable({
+                    _stream = stream,
+                }, stream_spaces_mt)
+            end
+            stream._gc = ffi.gc(ffi.new('char[1]'), function()
+                conn:on_schema_reload(nil, on_schema_reload)
+            end)
+            stream._conn:on_schema_reload(on_schema_reload)
```

Comment:
Can't we do without on_schema reload? Like this:

- Store a private table of all cached spaces and a schema version in each stream.
- When stream.space is accessed, compare stream._schema_version with the conn._schema_version.
- If they are the same, look up the space in the cache, otherwise cleanup the cache and reset stream._schema_version.
- If the space isn't found in the cache, look up the space in the conn.space table and store it in the cache.

Reply:
Your suggestion for spaces involves copying spaces from
connection to stream when creating a stream, this is a long process,
especially if streams are created very often. Now i implement lazy creation and updating of spaces.

Reply:
I think you misundrestood Vova. What he proposes is exactly what I proposed earlier - you create the space cache empty for each stream, like now. Then for each attempt to get a space you look up in the cache. If it is empty, you go to the connection's space cache. If found there, you wrap it into the stream's cache like it is done now. If not found, then there is no such space. The same for the indexes.

No entire schema copying done ever. And each space/index you wrap individually on demand. Just like now.

But additionally, when you create a stream, you remember in it the connection's schema version number. Then before you try to look up a space or an index in the stream's cache, you always first check if `stream.schema_version == connection.schema_version`. If they mismatch, you clear the space cache, remember the new version, and start from the scratch.

Reply:
Yes, that's exactly what I meant. However, if we set stream.__index to connection (as suggested in my other [comment](https://github.com/tarantool/tarantool/pull/6139/commits/76f1f89c44e509f5f3e67188d7362f753b71092d#r681706924)), we aren't going to need this anymore - stream.space will simply fall back on connection.space AFAIU.

Reply:
Lazy copying is still needed - stream.space.X needs to store a reference to stream (or stream_id) so that it can forward it to space and index methods.

Reply:
Fixed, we need chema_version comparison only for spaces, because all access for indexes is carried out throught space.